From a195976d635c3672cae684d6338655aa25f6d98c Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Tue, 6 Dec 2016 17:05:39 +0000
Subject: [PATCH] bcm2835-rng: Avoid initialising if already enabled

Avoids the 0x40000 cycles of warmup again if firmware has already used it
---
 drivers/char/hw_random/bcm2835-rng.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

--- a/drivers/char/hw_random/bcm2835-rng.c
+++ b/drivers/char/hw_random/bcm2835-rng.c
@@ -102,9 +102,10 @@ static int bcm2835_rng_probe(struct plat
 		rng_setup(rng_base);
 
 	/* set warm-up count & enable */
-	__raw_writel(RNG_WARMUP_COUNT, rng_base + RNG_STATUS);
-	__raw_writel(RNG_RBGEN, rng_base + RNG_CTRL);
-
+	if (!(__raw_readl(rng_base + RNG_CTRL) & RNG_RBGEN)) {
+		__raw_writel(RNG_WARMUP_COUNT, rng_base + RNG_STATUS);
+		__raw_writel(RNG_RBGEN, rng_base + RNG_CTRL);
+	}
 	/* register driver */
 	err = hwrng_register(&bcm2835_rng_ops);
 	if (err) {
